src/mlpack/methods/ann/image_functions/bilinear_function.hpp

```diff
@@ -0,0 +1,194 @@
+ */
+#ifndef MLPACK_METHODS_ANN_IMAGE_FUNCTIONS_BILINEAR_FUNCTION_HPP
+#define MLPACK_METHODS_ANN_IMAGE_FUNCTIONS_BILINEAR_FUNCTION_HPP
+
+#include <mlpack/prereqs.hpp>
+#include <assert.h>
+
+namespace mlpack {
+namespace ann /** Artificial Neural Network. */ {
+
+/**
+ * The bilinear interpolation function
+ *
+ */
+class BiLinearFunction
+{
+ /*
+  * The constructor for the bilinear interpolation.
+  * 
+  * @param inRowSize: Number of input rows
+  * @param inColSize: Number of input columns.
+  * @param outRowSize: Number of output rows.
+  * @param outColSie: Number of output columns.
+  */
+ public:
+  BiLinearFunction(
+      const size_t inRowSize,
+      const size_t inColSize,
+      const size_t outRowSize,
+      const size_t outColSize,
+      const size_t depth = 1):
+      inRowSize(inRowSize),
+      inColSize(inColSize),
+      outRowSize(outRowSize),
+      outColSize(outColSize),
+      depth(depth)
+  {};
+  /**
+   * UpSample the given input.
+   *
+   * @param input The input matrix
+   * @param output The resulting interpolated output image.
+   */
+  template<typename eT>
+  void UpSample(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+  {
+    if (output.is_empty())
+      output.set_size(outRowSize * outColSize, 1);
+
+    assert(output.n_rows == outRowSize * outColSize);
+    assert(output.n_cols == 1);
+
+    scaleRow = (double)inRowSize / (double) outRowSize;
+    scaleCol = (double)inColSize / (double) outColSize;
+
+    for (size_t k = 0; k < depth; k++)
+    {
+      for (size_t i = 0; i < outRowSize; i++)
+      {
+        for (size_t j = 0; j < outColSize; j++)
+        {
+          rOrigin = std::floor(i * scaleRow);
+          cOrigin = std::floor(j * scaleCol);
+
+          if (rOrigin > inRowSize - 2)
+            rOrigin = inRowSize - 2;
+          if (cOrigin > inColSize - 2)
+            cOrigin = inColSize - 2;
+
+          double deltaR = i * scaleRow - rOrigin;
+          double deltaC = j * scaleCol - cOrigin;
+          coeff1 = (1 - deltaR) * (1 - deltaC);
+          coeff2 = deltaR * (1 - deltaC);
+          coeff3 = (1 - deltaR) * deltaC;
+          coeff4 = deltaR * deltaC;
+
+          output(k * outRowSize * outColSize + j * outColSize + i) =
+              input(k* inRowSize * inColSize + cOrigin * inColSize + rOrigin) * coeff1 +
+              input(k* inRowSize * inColSize + cOrigin * inColSize + rOrigin + 1) * coeff2 +
+              input(k* inRowSize * inColSize + (cOrigin + 1) * inColSize + rOrigin) * coeff3 +
+              input(k* inRowSize * inColSize + (cOrigin + 1) * inColSize + rOrigin + 1) * coeff4;
+        }
+      }
+    }
+  }
+
+  /**
+   * DownSample the given input.
+   *
+   * @param input The input matrix
+   * @param output The resulting down-sampled output image.
+   */
+  template<typename eT>
+  void DownSample(const arma::Mat<eT>& input, arma::Mat<eT>& output)
+  {
+    if (output.is_empty())
+      output.set_size(inRowSize * inColSize, 1);
```

Comment:
That's incorrect. You should write
`output.set_size(inRowSize * inColSize * depth, 1);`

Reply:
Yup i changed it in the local branch. Will push it soon.